File: February 23/23 Feb/Unique Paths in a Grid gfg.cpp
    int uniquePaths(int n, int m, vector<vector<int>> &grid) {
        // code here
        vector<vector<long long>> paths(n,vector<long long>(m,0));
        
        long long mod=1e9+7;
        
        paths[n-1][m-1]=grid[n-1][m-1];
        
        for(int r=n-1;r>=0;r--)
        {
            for(int c=m-1;c>=0;c--)
            {
                if(r==(n-1) && c==(m-1)) 
                    continue;
                if(grid[r][c]==0)
                    continue;
                if((r+1)<n && grid[r+1][c]==1)
                    paths[r][c]=(paths[r][c] + paths[r+1][c])%mod;
                if((c+1)<m && grid[r][c+1]==1)
                    paths[r][c]=(paths[r][c] + paths[r][c+1])%mod;
            }
        }
        return paths[0][0];
    }

File: JUNE 2025/02 June/Unique Paths in a Grid gfg.java
class Solution {
    public int countUniquePaths(int[][] grid, int[][] dp, int r, int c, int N, int M)
    {
        if(!(r>=0 && r<N && c>=0 && c<M && grid[r][c]==0))
            return 0;

        if(dp[r][c]!=-1)
            return dp[r][c];
        
        int countPath = 0;
        
        countPath += countUniquePaths(grid, dp, r, c+1, N, M); // go right
        countPath += countUniquePaths(grid, dp, r+1, c, N, M); // go down
        
        return dp[r][c] = countPath;
    }
    
    public int uniquePaths(int[][] grid) {
        // code here
        int N = grid.length, M = grid[0].length;
        int[][] dp = new int[N+5][M+5];
        
        for(int[] row:dp)
            Arrays.fill(row, -1);
        
        if(grid[0][0]==1 || grid[N-1][M-1]==1)
            return 0;
        
        dp[N-1][M-1] = 1;
        
        return countUniquePaths(grid, dp, 0, 0, N, M);
    }
};
